applications/FluidDynamicsApplication/python_scripts/symbolic_generation/compressible_navier_stokes/templates/compressible_navier_stokes_explicit_template_2D4N.cpp

```diff
@@ -184,27 +222,36 @@ void CompressibleNavierStokesExplicit<2,4>::CalculateMomentumProjection(const Pr
     this->FillElementData(data, rCurrentProcessInfo);
 
     // Calculate shock capturing values
-    BoundedVector<double, Dim*NumNodes> mom_proj = ZeroVector(Dim*NumNodes);
+    constexpr std::size_t vector_size = Dim*NumNodes;
+    BoundedVector<double, vector_size> mom_proj = ZeroVector(vector_size);
 
     Vector N;
-    Matrix DN_DX_iso;
+    Matrix DN_De;
     Matrix DN_DX;
+
+    Matrix J;
     Matrix Jinv;
 
     auto& r_geometry = GetGeometry();
-    const auto& gauss_points = r_geometry.IntegrationPoints(GeometryData::IntegrationMethod::GI_GAUSS_2);
+    const auto& gauss_points = r_geometry.IntegrationPoints(GetIntegrationMethod());
     for(const auto& gauss_point: gauss_points)
     {
+        const double w = gauss_point.Weight();
         r_geometry.ShapeFunctionsValues(N, gauss_point.Coordinates());
-        r_geometry.InverseOfJacobian(Jinv, gauss_point.Coordinates());
-        r_geometry.ShapeFunctionsLocalGradients(DN_DX_iso, gauss_point.Coordinates());
-        GeometryUtils::ShapeFunctionsGradients(DN_DX_iso, Jinv, DN_DX);
+
+        double detJ;
+        r_geometry.Jacobian(J, gauss_point.Coordinates());
+        MathUtils<double>::InvertMatrix(J, Jinv, detJ);
```

Comment:
Same `InverseJacobian` comment.

Reply:
In this place in particular I need both the inverse and determinant, and `InverseOfJacobian` only yields the former.